LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DeviceLoaderSynchronizer.cs

```diff
@@ -304,26 +322,43 @@ private void LogRequestFailed(LoRaRequest request, LoRaDeviceRequestFailedReason
                 case LoRaDeviceRequestFailedReason.ApplicationError:
                     Logger.Log(deviceId, "problem resolving device", LogLevel.Error);
                     break;
+
+                case LoRaDeviceRequestFailedReason.BelongsToAnotherGateway:
+                    Logger.Log(deviceId, "device is not our device, ignore message", LogLevel.Debug);
+                    break;
             }
         }
 
         private async Task<LoRaDevice> InitializeDeviceAsync(LoRaDevice loRaDevice)
         {
             try
             {
+                // Our device if it does not have a gateway assigned or is assigned to our
+                var isOurDevice = string.IsNullOrEmpty(loRaDevice.GatewayID) || string.Equals(loRaDevice.GatewayID, this.configuration.GatewayID, StringComparison.InvariantCultureIgnoreCase);
+                // Only create client if the device is our
+                if (!isOurDevice)
+                {
+                    loRaDevice.IsOurDevice = false;
+                    return loRaDevice;
+                }
+
                 // Calling initialize async here to avoid making async calls in the concurrent dictionary
                 // Since only one device will be added, we guarantee that initialization only happens once
                 if (await loRaDevice.InitializeAsync())
                 {
+                    // revalidate based on device twin property
                     loRaDevice.IsOurDevice = string.IsNullOrEmpty(loRaDevice.GatewayID) || string.Equals(loRaDevice.GatewayID, this.configuration.GatewayID, StringComparison.InvariantCultureIgnoreCase);
```

Comment:
I assume we would not have to re-evaluate isOurDevice here, as we just did on line 337.

Reply:
This is necessary because InitializeAsync will read the twin which will update loRaDevice.GatewayID value. This address a problem where the Function cache is outdated.